pkg/ciliumenvoyconfig/exp_controller.go

```diff
@@ -0,0 +1,581 @@
+	LocalNodeStore *node.LocalNodeStore
+
+	NodeLabels    *nodeLabels
+	CECs          statedb.RWTable[*CEC]
+	EnvoySyncer   resourceMutator
+	PolicyTrigger policyTrigger
+}
+
+type resourceMutator interface {
+	DeleteEnvoyResources(context.Context, envoy.Resources) error
+	UpdateEnvoyResources(context.Context, envoy.Resources, envoy.Resources) error
+}
+
+type policyTrigger interface {
+	TriggerPolicyUpdates()
+}
+
+type cecController struct {
+	params cecControllerParams
+	writer *experimental.Writer
+}
+
+type cecControllerOut struct {
+	cell.Out
+
+	C    *cecController
+	Hook experimental.ServiceHook `group:"service-hooks"`
+}
+
+func newNodeLabels() *nodeLabels {
+	nl := &nodeLabels{}
+	lbls := map[string]string{}
+	nl.Store(&lbls)
+	return nl
+}
+
+func newCECController(params cecControllerParams) cecControllerOut {
+	if !params.ExpConfig.EnableExperimentalLB {
+		return cecControllerOut{}
+	}
+
+	c := &cecController{params, nil}
+	params.JobGroup.Add(job.OneShot("proxy-redirect-controller", c.proxyRedirectController))
+	params.JobGroup.Add(job.OneShot("backends-controller", c.backendController))
+	params.JobGroup.Add(job.OneShot("node-label-controller", c.nodeLabelController))
+	return cecControllerOut{
+		C:    c,
+		Hook: c.onServiceUpsert,
+	}
+}
+
+func (c *cecController) setWriter(w *experimental.Writer) {
+	if c != nil {
+		c.writer = w
+	}
+}
+
+// proxyRedirectController watches for changed CECs and updates the proxy redirect in services.
+func (c *cecController) proxyRedirectController(ctx context.Context, health cell.Health) error {
+	wtxn := c.params.DB.WriteTxn(c.params.CECs)
+	changeIter, err := c.params.CECs.Changes(wtxn)
+	wtxn.Commit()
+	if err != nil {
+		return err
+	}
+
+	svcs := c.writer.Services()
+	limiter := rate.NewLimiter(50*time.Millisecond, 3)
+	listeners := part.Set[uint16]{}
+
+	for {
+		wtxn := c.writer.WriteTxn()
+		changes, watch := changeIter.Next(wtxn)
+		oldListeners := listeners
+		for change := range changes {
+			cec := change.Object
+			if !change.Deleted && cec.Status.Kind != reconciler.StatusKindDone {
+				// Only process the CEC once it has been reconciled towards Envoy to avoid
+				// setting redirects before Envoy is ready.
+				continue
+			}
+
+			for _, port := range cec.Listeners.All() {
+				if change.Deleted {
+					listeners = listeners.Delete(port)
+				} else {
+					listeners = listeners.Set(port)
+				}
+			}
+
+			for _, svcl := range cec.Spec.Services {
+				svc, _, found := svcs.Get(wtxn, experimental.ServiceByName(svcl.ServiceName()))
+				if found && (change.Deleted || !getProxyRedirect(cec, svcl).Equal(svc.ProxyRedirect)) {
+					// Do an upsert to call into onServiceUpsert() to update the L7ProxyPort.
+					c.writer.UpsertService(wtxn, svc.Clone())
+				}
+			}
+		}
+		wtxn.Commit()
+
+		// When listeners change trigger the policy updates.
+		if !oldListeners.Equal(listeners) {
+			// TODO: Policy does not need to be recomputed for this, but if we do not 'force'
```

Comment:
I mean, policy does need to be recomputed; policy calculation includes L3/L4 (bpf) and L7 (Envoy / DNS) logic; there's no distinction.

We may wish to add some sort of pluggable predicate model in the future, but regenerating all endpoints is reasonably efficient and it's not like Envoy configs are high-churn.

Reply:
I guess this comment can stay for the time being; calculating policy and calculating proxy ports are separate steps.